src/domain/core/src/services/sync_service.rs

```diff
@@ -24,31 +25,54 @@ use crate::*;
 pub trait SyncService: Send + Sync {
     async fn sync(
         &self,
-        src: &DatasetRefAny,
-        dst: &DatasetRefAny,
+        request: SyncRequest,
         options: SyncOptions,
         listener: Option<Arc<dyn SyncListener>>,
-    ) -> Result<SyncResult, SyncError>;
-
-    async fn sync_multi(
-        &self,
-        requests: Vec<SyncRequest>,
-        options: SyncOptions,
-        listener: Option<Arc<dyn SyncMultiListener>>,
-    ) -> Vec<SyncResultMulti>;
+    ) -> Result<(SyncResult, Arc<dyn Dataset>), SyncError>;
 
     /// Adds dataset to IPFS and returns the root CID.
     /// Unlike `sync` it does not do IPNS resolution and publishing.
-    async fn ipfs_add(&self, src: &DatasetRef) -> Result<String, SyncError>;
+    async fn ipfs_add(&self, src: Arc<dyn Dataset>) -> Result<String, InternalError>;
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 pub struct SyncRequest {
-    pub src: DatasetRefAny,
-    pub dst: DatasetRefAny,
+    pub src: SyncRequestSource,
+    pub dst: SyncRequestDestination,
 }
 
-#[derive(Debug, Clone)]
+pub struct SyncRequestSource {
+    pub dataset: Arc<dyn Dataset>,
+    pub src_ref: DatasetRefAny,
+    pub sync_ref: SyncRef,
```

Comment:
This and `SyncRequestDestination` structs stand out to me compared to the simplicity of ingest/transform services.

I kinda expected something like:
```rust
pub struct SyncRequest {
    pub src: SyncRef,
    pub dst: SyncRef,
}

pub enum SyncRef {
  Local(ResolvedDataset),
  LocalNew(DatasetAlias),
  Remote(Arc<Url>),
}
```
and have a clear separation of layer that resolves `DatasetRefAny` into one of these `SyncRef`s (`SyncPlanner`?) so that `DatasetRef`s were not appearing in the low-level sync API just like they don't appear in pull plans.

Just looking at this struct I'm not sure what to expect from `src_ref` and `sync_ref` and how they overlap. Some comments would help.

Reply:
Almost.

```
pub struct SyncRequest {
    pub src: SyncRef,
    pub dst: SyncRef,
}

pub enum SyncRef {
    Local(ResolvedDataset),
    LocalNew(DatasetAlias),
    Remote(SyncRefRemote),
}

pub struct SyncRefRemote {
    pub url: Arc<Url>,
    pub dataset: Arc<dyn Dataset>,
    pub original_remote_ref: DatasetRefRemote,
}
```

The original remote URL is important for progress reporting and some tests.